tests/integration/test_charm.py

```diff
@@ -37,24 +42,248 @@ async def test_build_and_deploy(ops_test: OpsTest):
+    )
+    primary_unit_name = primary_unit.name
+
+    assert len(ops_test.model.applications[APP_NAME].units) == 3
+
+    # Destroy the primary unit and wait 5 seconds to ensure that the
+    # juju status changed from active
+    await ops_test.model.destroy_units(primary_unit.name)
+    time.sleep(5)
+    await ops_test.model.wait_for_idle(
+        apps=[APP_NAME],
+        status="active",
+        raise_on_blocked=True,
+        timeout=1000,
+    )
+
+    # Wait for unit to be destroyed and confirm that the new primary unit is different
+    assert len(ops_test.model.applications[APP_NAME].units) == 2
+
+    random_unit = ops_test.model.applications[APP_NAME].units[0]
+    new_primary_unit = await get_primary_unit(
+        ops_test,
+        random_unit,
+        server_config_credentials["username"],
+        server_config_credentials["password"],
+    )
+
+    assert primary_unit_name != new_primary_unit.name
+
+    # Add the unit back and wait until it is active
+    await ops_test.model.applications[APP_NAME].add_unit()
+    time.sleep(5)
+    await ops_test.model.wait_for_idle(
+        apps=[APP_NAME],
+        status="active",
+        raise_on_blocked=True,
+        timeout=1000,
+    )
+
+    assert len(ops_test.model.applications[APP_NAME].units) == 3
+
+
+@pytest.mark.order(4)
+@pytest.mark.abort_on_fail
+async def test_cluster_preserves_data_on_delete(ops_test: OpsTest):
+    """Test that data is preserved during scale up and scale down."""
+    # Insert values into test table from the primary unit
+    random_unit = ops_test.model.applications[APP_NAME].units[0]
+    server_config_credentials = await get_server_config_credentials(random_unit)
+
+    primary_unit = await get_primary_unit(
+        ops_test,
+        random_unit,
+        server_config_credentials["username"],
+        server_config_credentials["password"],
+    )
+
+    random_chars = generate_random_string(40)
+    create_records_sql = (
+        "CREATE DATABASE IF NOT EXISTS test",
+        "CREATE TABLE IF NOT EXISTS test.instance_state_replication (id varchar(40), primary key(id))",
+        f"INSERT INTO test.instance_state_replication VALUES ('{random_chars}')",
+    )
+
+    mysqlsh_sql_base_command = f"mysqlsh --sql {server_config_credentials['username']}:{server_config_credentials['password']}@127.0.0.1"
+
+    await run_command_on_unit(
+        primary_unit, f"{mysqlsh_sql_base_command} -e \"{';'.join(create_records_sql)}\""
+    )
+
+    assert len(ops_test.model.applications[APP_NAME].units) == 3
+    old_unit_names = [unit.name for unit in ops_test.model.applications[APP_NAME].units]
+
+    # Add a unit and wait until it is active
+    await ops_test.model.applications[APP_NAME].add_unit()
+    time.sleep(5)
+    await ops_test.model.wait_for_idle(
+        apps=[APP_NAME],
+        status="active",
+        raise_on_blocked=True,
+        timeout=1000,
+    )
+
+    assert len(ops_test.model.applications[APP_NAME].units) == 4
```

Comment:
This is an opinion, and applies to other tests above. 
Are these `assert`ions part of the test, or are they just 'if' statements?

If the test fails because there weren't 4 units for some unknown reason, it doesn't mean that `cluster_preserves_data_on_delete` wasn't met, which would make debugging more complicated. Maybe make these if-statements, and fail the test with a helpful message. Same to the tests above. We should probably keep assertions to failure conditions for the 'thing' we're testing, not for validating set-up.

Again, it's possible I am misunderstanding.

Reply:
Removed assertions in ff877d16df482a9781067a33b0a1cf1e46591782 (where I am only testing the 'thing' to be tested)